File: applications/faas-migration-go/test_api.py
import sys
import requests
import random

NotFoundTestGuid = "8324cf72-84c6-4001-bec2-46505dbe4301"
RandomStringCharset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"


def GenerateRandomString(length):
    return ''.join(random.choice(RandomStringCharset) for _ in range(length))


class InsertRequest:
    def __init__(self, title, description):
        self.title = title
        self.description = description

    @classmethod
    def fromjson(cls, json_object):
        return cls(json_object['title'], json_object['description'])

    def tojson(self):
        return {'title': self.title, 'description': self.description}


class ToDoItem:
    def __init__(self, item_id, title, description, insertion_timestamp, done_timestamp):
        self.id = item_id
        self.title = title
        self.description = description
        self.insertion_timestamp = insertion_timestamp
        self.done_timestamp = done_timestamp

    @classmethod
    def fromjson(cls, json_object):
        return cls(json_object['ID'], json_object['title'], json_object['description'],
                   json_object['insertion_timestamp'], json_object['done_timestamp'])

    def tojson(self):
        return {'ID': self.id, 'title': self.title, 'description': self.description,
                'insertion_timestamp': self.insertion_timestamp, 'done_timestamp': self.done_timestamp}


class API:
    def __init__(self, endpoint):
        if endpoint[-1] == '/':
            self.endpoint = endpoint
        else:
            self.endpoint = endpoint + "/"

    # API functions

    def listItems(self):
        resp = requests.get(url=self.endpoint + "lst")
        ret = []
        if resp.status_code == 200:
            if resp.content is not None:
                data = resp.json()
                if type(data) == list:
                    for item in data:
                        ret.append(ToDoItem.fromjson(item))
                else:
                    ret.append(ToDoItem.fromjson(data))
        else:
            print("failed listItems() with error code " + str(resp.status_code), file=sys.stderr)
            return None
        return ret

    def instertItem(self, insert_request):
        resp = requests.post(url=self.endpoint + "put", json=insert_request.tojson())
        if resp.status_code == 200:
            data = resp.json()
            return ToDoItem.fromjson(data)
        else:
            print("failed insertItem() with error code " + str(resp.status_code), file=sys.stderr)
            return None

    def markAsDone(self, item_id):
        resp = requests.post(url=self.endpoint + "done?id=" + item_id)
        if resp.status_code != 200:
            print("failed markAsDone() with error code " + str(resp.status_code), file=sys.stderr)
            return None
        return item_id

    def deleteItem(self, item_id):
        resp = requests.post(url=self.endpoint + "del?id=" + item_id)
        if resp.status_code != 200:
            print("failed deleteItem() with error code " + str(resp.status_code), file=sys.stderr)
            return None
        return item_id

    def getItem(self, item_id):
        resp = requests.get(url=self.endpoint + "get?id=" + item_id)
        if resp.status_code == 200:
            data = resp.json()
            return ToDoItem.fromjson(data)
        else:
            print("failed insertItem() with error code " + str(resp.status_code), file=sys.stderr)
            return None

    # Tests

    def checkAvailablity(self):
        res = self.listItems() is not None
        if res:
            print("API available!")
        else:
            print("API unavailable!")
        return res

    def insertItems(self, count):
        print("Inserting %d ToDo Items..." % count)
        insert_requests = []
        items = []
        for i in range(count):
            ir = InsertRequest("Todo-Item-%d" % i, GenerateRandomString(300))
            insert_requests.append(ir)
            item = self.instertItem(ir)
            if item is None:
                return None
            items.append(item)
        for i in range(len(insert_requests)):
            if insert_requests[i].title != items[i].title:
                print("failed insertItems() with a unequal titles!", file=sys.stderr)
                return None
            elif insert_requests[i].description != items[i].description:
                print("failed insertItems() with a unequal descriptions!", file=sys.stderr)
                return None
        return items

    def checkListItems(self, items, mustExist, checkIsDone):
        print("Requesting List of Items...")
        itemsFromApp = self.listItems()
        if items is None:
            return None
        for i in range(len(items)):
            found = False
            for j in range(len(itemsFromApp)):
                if items[i].title == itemsFromApp[j].title and items[i].description == itemsFromApp[j].description:
                    found = True
                    if checkIsDone and itemsFromApp[j].done_timestamp == -1:
                        print("Fail!\nFound item that is expected to be Done but is not!", file=sys.stderr)
                        return None
                    break
            if not found and mustExist:
                print("Fail!\nCould not find an Item!", file=sys.stderr)
                return None
            if found and not mustExist:
                print("Fail!\nFound Item that should be deleted!!", file=sys.stderr)
                return None
        print("Done!")
        return "ok!"

    def markItemsAsDone(self, items):
        print("Marking %d items as Done..." % len(items))
        for i in range(len(items)):
            if self.markAsDone(items[i].id) is None:
                return None
        print("Done!")
        return "ok!"

    def checkItemsNotDone(self, items):
        print("Using 'Get' Function to check if the %d items are Done..." % len(items))
        for i in range(len(items)):
            item = self.getItem(items[i].id)
            if item is None:
                return None
            if item.id != items[i].id:
                print("Fail!\nGet Item returned a different ID then queried.", file=sys.stderr)
                return None
            if item.done_timestamp != -1:
                print("Fail!\nAn Item is marked as done while it should not!", file=sys.stderr)
                return None
        print("Done!")
        return "ok!"

    def checkDelete(self, items):
        print("Deleting all inserted Items...")
        for item in items:
            if self.deleteItem(item.id) is None:
                return None
        print("Done!")
        return "ok!"

    def checkFunctionsNotFoundBehaviour(self):
        print("Checking ID specific functions API Behaviour if ID does not exist")
        print("Get Function...")
        if self.getItem(NotFoundTestGuid) is not None:
            print("Fail!\nExpected Error. Got None.", file=sys.stderr)
            return None
        print("Ok! Got error while getting Item.\nDone Function...")
        if self.markAsDone(NotFoundTestGuid) is not None:
            print("Fail!\nExpected Error. Got None.", file=sys.stderr)
            return None
        print("Ok! Got error while marking as done.\nDelete Function...")
        if self.deleteItem(NotFoundTestGuid) is not None:
            print("Fail!\nExpected Error. Got None.", file=sys.stderr)
            return None
        print("Ok Got error while deleting item!\nDone!")
        return "ok!"


def test_api(endpoint, count):
    api = API(endpoint)
    if not api.checkAvailablity():
        return
    items = api.insertItems(count)
    if items is None:
        return
    if api.checkListItems(items, True, False) is None:
        return
    bound = int(count/2)
    print("Using Bound %d" % bound)
    doneItems = items[:bound]
    notDoneItems = items[bound:]
    if api.markItemsAsDone(doneItems) is None:
        return
    if api.checkListItems(doneItems, True, True) is None:
        return
    if api.checkItemsNotDone(notDoneItems) is None:
        return
    if api.checkDelete(items) is None:
        return
    if api.checkListItems(items, False, False) is None:
        return
    # if api.checkFunctionsNotFoundBehaviour() is None:
    #     return
    print("Success! All Checks have passed!")
    return "ok!"

File: applications/faas-migration-go/test_api.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"strings"
	"time"
)

// -----------------------------------------------------------------------------
// Input Configuration
// -----------------------------------------------------------------------------

var (
	endpoint = flag.String("endpoint", "undefined", "The endpoint to the ToDo API implementation")
	count    = flag.Int("count", 50, "The number of items to insert.")
)

func init() {
	flag.Parse()
}

// -----------------------------------------------------------------------------
// Model Definitions
// -----------------------------------------------------------------------------

// This type defines the input object used to create a ToDo Item
type InsertRequest struct {
	Title       string `json:"title"`
	Description string `json:"description"`
}
type ToDoItem struct {
	ID                 string `json:"ID"`
	Title              string `json:"title"`
	Description        string `json:"description"`
	InsertionTimestamp int64  `json:"insertion_timestamp"`
	DoneTimestamp      int64  `json:"done_timestamp"`
}

// -----------------------------------------------------------------------------
// API Interactions
// -----------------------------------------------------------------------------

type api string

// Request all items from the api
func (endpoint api) ListItems() ([]ToDoItem, error) {
	res, err := http.Get(fmt.Sprintf("%slst", endpoint))
	if err != nil {
		return nil, err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	data, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var output []ToDoItem
	err = json.Unmarshal(data, &output)
	if err != nil {
		return nil, fmt.Errorf("Contents: %s Original Error: %s", string(data), err.Error())
	}

	return output, nil
}

func (endpoint api) InsertItem(i InsertRequest) (*ToDoItem, error) {
	data, err := json.Marshal(i)
	if err != nil {
		return nil, err
	}

	res, err := http.Post(fmt.Sprintf("%sput", endpoint), "application/json", bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	// Best Practice: Close body when Function is done...
	defer res.Body.Close()

	resdata, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	var item ToDoItem
	err = json.Unmarshal(resdata, &item)
	if err != nil {
		return nil, fmt.Errorf("Contents: %s Original Error: %s", string(data), err.Error())
	}

	return &item, nil
}

func (endpoint api) MarkAsDone(id string) error {
	res, err := http.Post(fmt.Sprintf("%sdone?id=%s", endpoint, id), "application/json", bytes.NewReader([]byte{}))
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode == 200 {
		return nil
	}
	return errors.New(fmt.Sprintf("API returned status code: %d instead of 200", res.StatusCode))
}
func (endpoint api) DeleteItem(id string) error {
	res, err := http.Post(fmt.Sprintf("%sdel?id=%s", endpoint, id), "application/json", bytes.NewReader([]byte{}))
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode == 200 {
		return nil
	}
	return errors.New(fmt.Sprintf("API returned status code: %d instead of 200", res.StatusCode))
}

func (endpoint api) GetItem(s string) (*ToDoItem, error) {
	res, err := http.Get(fmt.Sprintf("%sget?id=%s", endpoint, s))
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	if res.StatusCode != 200 {
		return nil, errors.New(fmt.Sprintf("API returned status code: %d instead of 200", res.StatusCode))
	}
	data, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var item ToDoItem
	err = json.Unmarshal(data, &item)
	if err != nil {
		return nil, fmt.Errorf("Contents: %s Original Error: %s", string(data), err.Error())
	}
	return &item, nil
}

// -----------------------------------------------------------------------------
// Main Method
// -----------------------------------------------------------------------------

func main() {
	appUrl := *endpoint
	if appUrl == "undefined" {
		fmt.Println("The -endpoint Flag is required!")
		os.Exit(1)
	}
	// Append Trailing / if it is missing
	if !strings.HasSuffix(appUrl, "/") {
		appUrl += "/"
	}
	api := api(appUrl)
	api.CheckAvailability()
	items := api.InsertItems()
	api.CheckListItems(items, true, false)
	bound := (*count)/2
	fmt.Printf("Using Bound %d\n",bound)
	doneItems := items[:bound]
	notDoneItems := items[bound:]
	api.MarkItemsAsDone(doneItems)
	api.CheckListItems(doneItems, true, true)
	api.CheckItemsNotDone(notDoneItems)
	api.CheckDelete(items)
	api.CheckListItems(items, false, false)
	api.CheckFunctionsNotFoundBehaviour()
	fmt.Println("Success! All Checks have passed!")
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

func (endpoint api) CheckAvailability() {
	// Call list to check if API is up
	fmt.Printf("Checking if API is Available....")
	_, err := endpoint.ListItems()
	if err != nil {
		fmt.Printf("Fail!\n")
		fmt.Printf("Ping API by calling List has Failed!\nError: %s\n", err.Error())
		os.Exit(1)
	}
	fmt.Printf("Ok!\n")
}

func (endpoint api) InsertItems() []ToDoItem {
	fmt.Printf("Inserting %d ToDo Items...", *count)
	insertionRequests := make([]InsertRequest, 0)
	items := make([]ToDoItem, 0)
	for i := 0; i < *count; i++ {
		ireq := InsertRequest{
			Title:       fmt.Sprintf("Todo-Item-#%d", i),
			Description: GenerateRandomString(300),
		}
		insertionRequests = append(insertionRequests, ireq)

		item, err := endpoint.InsertItem(ireq)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
		items = append(items, *item)
	}
	fmt.Printf("Done!\nValidating Responses....")
	if len(items) != len(insertionRequests) {
		fmt.Printf("Fail!\nResponse Count is not equal to Request Count\n")
		os.Exit(1)
	}
	for k, req := range insertionRequests {
		res := items[k]
		if res.Title != req.Title {
			if len(items) != len(insertionRequests) {
				fmt.Printf("Fail!\nA Title Did not match\n")
				os.Exit(1)
			}
		}
		if res.Description != req.Description {
			if len(items) != len(insertionRequests) {
				fmt.Printf("Fail!\nA Description Did not match\n")
				os.Exit(1)
			}
		}
	}
	fmt.Printf("Done!\n")
	return items
}

func (endpoint api) CheckListItems(items []ToDoItem, mustExist bool, checkIsDone bool) {
	fmt.Printf("Requesting List of Items...")
	itemsFromApp, err := endpoint.ListItems()
	if err != nil {
		fmt.Printf("Fail!\nError: %s\n", err.Error())
		os.Exit(1)
	}
	fmt.Printf("Done!\nSearching for inserted items...")
	for _, item := range items {
		found := false
		for _, ifa := range itemsFromApp {
			if ifa.ID == item.ID && ifa.Title == item.Title &&
				ifa.Description == item.Description {
				found = true
				if checkIsDone && ifa.DoneTimestamp == -1 {
					fmt.Printf("Fail!\nFound item that is expected to be Done but is not!\n")
					os.Exit(1)
				}
				break
			}
		}
		if !found && mustExist {
			fmt.Printf("Fail!\nCould not find an Item!\n")
			os.Exit(1)
		}
		if found && !mustExist {
			fmt.Printf("Fail!\nFound Item that should be deleted!!\n")
			os.Exit(1)
		}
	}
	fmt.Printf("Done!\n")
}

func (endpoint api) MarkItemsAsDone(items []ToDoItem) {
	fmt.Printf("Marking %d items as Done...", len(items))

	for _, item := range items {
		err := endpoint.MarkAsDone(item.ID)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
	}

	fmt.Printf("Done!\n")
}

func (endpoint api) CheckItemsNotDone(items []ToDoItem) {
	fmt.Printf("Using 'Get' Function to check if the %d items are Done...", len(items))

	for _, item := range items {
		i, err := endpoint.GetItem(item.ID)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
		if i.ID != item.ID {
			fmt.Printf("Fail!\nGet Item returned a different ID then queried.\n")
			os.Exit(1)
		}
		if i.DoneTimestamp != -1 {
			fmt.Printf("Fail!\nAn Item is marked as done while it should not!\n")
			os.Exit(1)
		}
	}

	fmt.Printf("Done!\n")
}

func (endpoint api) CheckDelete(items []ToDoItem) {
	fmt.Printf("Deleting all inserted Items...")
	for _, item := range items {
		err := endpoint.DeleteItem(item.ID)
		if err != nil {
			fmt.Printf("Fail!\nError: %s\n", err.Error())
			os.Exit(1)
		}
	}
	fmt.Printf("Done!\n")
}

func (endpoint api) CheckFunctionsNotFoundBehaviour() {
	fmt.Println("Checking ID specific functions API Behaviour if ID does not exist")
	fmt.Printf("Get Function...")
	_, err := endpoint.GetItem(NotFoundTestGuid)
	if err == nil {
		fmt.Printf("Fail!\nExpected Error. Got None.\n")
		os.Exit(1)
	}
	fmt.Printf("Ok! Got error: %s\nDone Function...", err.Error())
	err = endpoint.MarkAsDone(NotFoundTestGuid)
	if err == nil {
		fmt.Printf("Fail!\nExpected Error. Got None.\n")
		os.Exit(1)
	}
	fmt.Printf("Ok! Got error: %s\nDelete Function...", err.Error())
	err = endpoint.DeleteItem(NotFoundTestGuid)
	if err == nil {
		fmt.Printf("Fail!\nExpected Error. Got None.\n")
		os.Exit(1)
	}
	fmt.Printf("Ok Got error: %s!\n", err.Error())
}

// -----------------------------------------------------------------------------
// Utility Methods
// -----------------------------------------------------------------------------

const NotFoundTestGuid = "8324cf72-84c6-4001-bec2-46505dbe4301"

const randomStringCharset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"

var charsetRunes []rune

func init() {
	charsetRunes = bytes.Runes([]byte(randomStringCharset))
	rand.Seed(time.Now().Unix())
}

func GenerateRandomString(l int) string {
	id := ""
	for i := 0; i < l; i++ {
		idx := rand.Intn(len(charsetRunes))
		id += string(charsetRunes[idx])
	}
	return id
}
